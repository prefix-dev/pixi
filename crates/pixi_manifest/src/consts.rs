use url::Url;

pub const PYPI_DEPENDENCIES: &str = "pypi-dependencies";
pub const DEFAULT_ENVIRONMENT_NAME: &str = "default";
pub const PROJECT_MANIFEST: &str = "pixi.toml";
pub const PYPROJECT_MANIFEST: &str = "pyproject.toml";
pub const DEFAULT_FEATURE_NAME: &str = DEFAULT_ENVIRONMENT_NAME;
pub const PYPROJECT_PIXI_PREFIX: &str = "tool.pixi";

lazy_static::lazy_static! {
    pub static ref DEFAULT_PYPI_INDEX_URL: Url = Url::parse("https://pypi.org/simple").unwrap();
}
