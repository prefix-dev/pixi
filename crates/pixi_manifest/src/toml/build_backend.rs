use std::collections::BTreeMap;

use indexmap::IndexMap;
use pixi_spec::TomlSpec;
use pixi_toml::{Same, TomlFromStr, TomlIndexMap, TomlWith};
use rattler_conda_types::NamedChannelOrUrl;
use toml_span::{DeserError, Spanned, Value, de_helpers::TableHelper, value::ValueInner};

use crate::{
    PackageBuild, TargetSelector, TomlError,
    build_system::BuildBackend,
    error::GenericError,
    toml::build_target::TomlPackageBuildTarget,
    utils::{PixiSpanned, package_map::UniquePackageMap},
};

#[derive(Debug)]
pub struct TomlPackageBuild {
    pub backend: PixiSpanned<TomlBuildBackend>,
    pub channels: Option<PixiSpanned<Vec<NamedChannelOrUrl>>>,
    pub additional_dependencies: UniquePackageMap,
    pub configuration: Option<serde_value::Value>,
    pub target: IndexMap<PixiSpanned<TargetSelector>, TomlPackageBuildTarget>,
}

#[derive(Debug)]
pub struct TomlBuildBackend {
    pub name: PixiSpanned<rattler_conda_types::PackageName>,
    pub spec: TomlSpec,
}

impl TomlPackageBuild {
    pub fn into_build_system(self) -> Result<PackageBuild, TomlError> {
        // Parse the build backend and ensure it is a binary spec.
        let build_backend_spec = self.backend.value.spec.into_spec().map_err(|e| {
            TomlError::Generic(
                GenericError::new(e.to_string()).with_opt_span(self.backend.span.clone()),
            )
        })?;

        // Convert the additional dependencies and make sure that they are binary.
        let additional_dependencies = self.additional_dependencies.specs;

        // Make sure there are no empty channels
        if let Some(channels) = &self.channels {
            if channels.value.is_empty() {
                return Err(TomlError::Generic(
                    GenericError::new("No channels specified for the build backend dependencies")
                        .with_opt_span(channels.span()),
                ));
            }
        }

        // Convert target-specific build configurations
        let target_configuration = self
            .target
            .into_iter()
            .flat_map(|(selector, target)| {
                target
                    .configuration
                    .map(|config| (selector.into_inner(), config))
            })
            .collect::<IndexMap<_, _>>();

        Ok(PackageBuild {
            backend: BuildBackend {
                name: self.backend.value.name.value,
                spec: build_backend_spec,
            },
            additional_dependencies,
            channels: self.channels.map(|channels| channels.value),
            configuration: self.configuration,
            target_configuration: if target_configuration.is_empty() {
                None
            } else {
                Some(target_configuration)
            },
        })
    }
}

pub fn convert_toml_to_serde(value: &mut Value) -> Result<serde_value::Value, DeserError> {
    Ok(match value.take() {
        ValueInner::String(s) => serde_value::Value::String(s.to_string()),
        ValueInner::Integer(i) => serde_value::Value::I64(i),
        ValueInner::Float(f) => serde_value::Value::F64(f),
        ValueInner::Boolean(b) => serde_value::Value::Bool(b),
        ValueInner::Array(mut arr) => {
            let mut json_arr = Vec::new();
            for item in &mut arr {
                json_arr.push(convert_toml_to_serde(item)?);
            }
            serde_value::Value::Seq(json_arr)
        }
        ValueInner::Table(table) => {
            let mut map = BTreeMap::new();
            for (key, mut val) in table {
                map.insert(
                    serde_value::Value::String(key.to_string()),
                    convert_toml_to_serde(&mut val)?,
                );
            }
            serde_value::Value::Map(map)
        }
    })
}

impl<'de> toml_span::Deserialize<'de> for TomlBuildBackend {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        let mut th = TableHelper::new(value)?;
        let name = th.required_s::<TomlFromStr<rattler_conda_types::PackageName>>("name")?;
        th.finalize(Some(value))?;

        let spec = toml_span::Deserialize::deserialize(value)?;

        Ok(TomlBuildBackend {
            name: PixiSpanned::from(Spanned {
                value: name.value.into_inner(),
                span: name.span,
            }),
            spec,
        })
    }
}

impl<'de> toml_span::Deserialize<'de> for TomlPackageBuild {
    fn deserialize(value: &mut Value<'de>) -> Result<Self, DeserError> {
        let mut th = TableHelper::new(value)?;
        let build_backend = th.required_s("backend")?.into();
        let channels = th
            .optional_s::<TomlWith<_, Vec<TomlFromStr<_>>>>("channels")
            .map(|s| PixiSpanned {
                value: s.value.into_inner(),
                span: Some(s.span.start..s.span.end),
            });
        let additional_dependencies = th.optional("additional-dependencies").unwrap_or_default();

        let configuration = th
            .take("configuration")
            .map(|(_, mut value)| convert_toml_to_serde(&mut value))
            .transpose()?;

        let target = th
            .optional::<TomlWith<_, TomlIndexMap<_, Same>>>("target")
            .map(TomlWith::into_inner)
            .unwrap_or_default();

        th.finalize(None)?;
        Ok(Self {
            backend: build_backend,
            channels,
            additional_dependencies,
            configuration,
            target,
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;
    use pixi_test_utils::format_parse_error;

    use super::*;

    fn expect_parse_failure(pixi_toml: &str) -> String {
        let parse_error = <TomlPackageBuild as crate::toml::FromTomlStr>::from_toml_str(pixi_toml)
            .and_then(TomlPackageBuild::into_build_system)
            .expect_err("parsing should fail");

        format_parse_error(pixi_toml, parse_error)
    }

    #[test]
    fn test_configuration_parsing() {
        let toml = r#"
            backend = { name = "foobar", version = "*" }
            configuration = { key = "value", other = ["foo", "bar"], integer = 1234, nested = { abc = "def" } }
        "#;
        let parsed = <TomlPackageBuild as crate::toml::FromTomlStr>::from_toml_str(toml)
            .expect("parsing should succeed");

        insta::assert_debug_snapshot!(parsed);
    }

    #[test]
    fn test_missing_version_specifier() {
        assert_snapshot!(expect_parse_failure(
            r#"
            backend = { name = "foobar" }
        "#
        ));
    }

    #[test]
    fn test_missing_backend() {
        assert_snapshot!(expect_parse_failure(""));
    }

    #[test]
    fn test_missing_name() {
        assert_snapshot!(expect_parse_failure(
            r#"
            backend = { version = "12.*" }
        "#
        ));
    }

    #[test]
    fn test_empty_channels() {
        assert_snapshot!(expect_parse_failure(
            r#"
            backend = { name = "foobar", version = "*" }
            channels = []
        "#
        ));
    }

    #[test]
    fn test_additional_build_backend_keys() {
        assert_snapshot!(expect_parse_failure(
            r#"
            backend = { name = "foobar", version = "*", sub = "bar" }
        "#
        ));
    }

    #[test]
    fn test_additional_keys() {
        assert_snapshot!(expect_parse_failure(
            r#"
            backend = { name = "foobar", version = "*" }
            additional = "key"
        "#
        ));
    }
}
