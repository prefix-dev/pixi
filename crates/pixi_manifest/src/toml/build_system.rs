use itertools::Either;
use pixi_spec::TomlSpec;
use rattler_conda_types::NamedChannelOrUrl;
use serde::Deserialize;

use crate::{
    build_system::BuildBackend,
    utils::{package_map::UniquePackageMap, PixiSpanned},
    BuildSystem, TomlError,
};

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct TomlBuildSystem {
    pub build_backend: PixiSpanned<TomlBuildBackend>,

    #[serde(default)]
    pub channels: Option<PixiSpanned<Vec<NamedChannelOrUrl>>>,

    #[serde(default)]
    pub additional_dependencies: UniquePackageMap,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct TomlBuildBackend {
    pub name: PixiSpanned<rattler_conda_types::PackageName>,

    #[serde(flatten)]
    pub spec: TomlSpec,
}

impl TomlBuildSystem {
    pub fn from_toml_str(source: &str) -> Result<Self, TomlError> {
        toml_edit::de::from_str(source).map_err(TomlError::from)
    }

    pub fn into_build_system(self) -> Result<BuildSystem, TomlError> {
        // Parse the build backend and ensure it is a binary spec.
        let build_backend_spec = self
            .build_backend
            .value
            .spec
            .into_binary_spec()
            .map_err(|e| {
                TomlError::Generic(e.to_string().into(), self.build_backend.span.clone())
            })?;

        // Convert the additional dependencies and make sure that they are binary.
        let additional_dependencies = self
            .additional_dependencies
            .specs
            .into_iter()
            .map(|(name, spec)| match spec.into_source_or_binary() {
                Either::Right(binary) => Ok((name, binary)),
                Either::Left(_source) => {
                    let spec_range = self
                        .additional_dependencies
                        .value_spans
                        .get(&name)
                        .or_else(|| self.additional_dependencies.name_spans.get(&name))
                        .cloned();
                    Err(TomlError::Generic(
                        "Cannot use source dependencies for build backends dependencies".into(),
                        spec_range,
                    ))
                }
            })
            .collect::<Result<_, TomlError>>()?;

        // Make sure there are no empty channels
        if let Some(channels) = &self.channels {
            if channels.value.is_empty() {
                return Err(TomlError::Generic(
                    "No channels specified for the build backend dependencies".into(),
                    channels.span(),
                ));
            }
        }

        Ok(BuildSystem {
            build_backend: BuildBackend {
                name: self.build_backend.value.name.value,
                spec: build_backend_spec,
                additional_args: None,
            },
            additional_dependencies,
            channels: self.channels.map(|channels| channels.value),
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;

    use super::*;
    use crate::utils::test_utils::format_parse_error;

    fn expect_parse_failure(pixi_toml: &str) -> String {
        let parse_error = TomlBuildSystem::from_toml_str(pixi_toml)
            .and_then(TomlBuildSystem::into_build_system)
            .expect_err("parsing should fail");

        format_parse_error(pixi_toml, parse_error)
    }

    #[test]
    fn test_disallow_source() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", git = "https://github.com/org/repo" }
        "#
        ));
    }

    #[test]
    fn test_missing_version_specifier() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar" }
        "#
        ));
    }

    #[test]
    fn test_missing_backend() {
        assert_snapshot!(expect_parse_failure(""));
    }

    #[test]
    fn test_missing_name() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { version = "12.*" }
        "#
        ));
    }

    #[test]
    fn test_empty_channels() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", version = "*" }
            channels = []
        "#
        ));
    }

    #[test]
    fn test_additional_build_backend_keys() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", version = "*", foo = "bar" }
        "#
        ));
    }

    #[test]
    fn test_additional_keys() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", version = "*" }
            additional = "key"
        "#
        ));
    }
}
