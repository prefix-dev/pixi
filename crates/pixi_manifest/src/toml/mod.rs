mod build_system;
mod channel;
mod document;
mod environment;
mod feature;
mod manifest;
mod package;
mod target;
mod workspace;

pub use build_system::TomlBuildSystem;
pub use channel::TomlPrioritizedChannel;
pub use document::TomlDocument;
pub use environment::{TomlEnvironment, TomlEnvironmentList};
pub use feature::TomlFeature;
pub use manifest::TomlManifest;
pub use package::{ExternalPackageProperties, PackageError, TomlPackage};
pub use target::TomlTarget;
pub use workspace::{ExternalWorkspaceProperties, TomlWorkspace, WorkspaceError};
