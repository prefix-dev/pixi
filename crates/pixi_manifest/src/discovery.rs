//! This module provides the [`WorkspaceDiscoverer`] struct which can be used to
//! discover the workspace manifest in a directory tree.

use std::{
    path::{Path, PathBuf},
    sync::Arc,
};

use miette::{Diagnostic, NamedSource};
use pixi_consts::consts;
use rattler_conda_types::Version;
use thiserror::Error;
use toml_span::Deserialize;

use crate::toml::ExternalWorkspaceProperties;
use crate::{
    pyproject::PyProjectManifest,
    toml::{ExternalPackageProperties, TomlManifest},
    utils::WithSourceCode,
    warning::WarningWithSource,
    AssociateProvenance, ManifestKind, ManifestProvenance, ManifestSource, PackageManifest,
    ProvenanceError, TomlError, WithProvenance, WithWarnings, WorkspaceManifest,
};

/// A helper struct to discover the workspace manifest in a directory tree from
/// a given path. The discoverer will walk up the directory tree until it finds
/// a workspace.
///
/// It can also collect the first package that was found on the way to the
/// workspace. See [`WorkspaceDiscoverer::with_closest_package`] for more
/// information.
pub struct WorkspaceDiscoverer {
    /// The current path
    start: DiscoveryStart,

    /// Also discover the package closest to the current directory.
    discover_package: bool,
}

/// A workspace discovered by calling [`WorkspaceDiscoverer::discover`].
#[derive(Debug)]
pub struct Manifests {
    /// The discovered workspace manifest
    pub workspace: WithProvenance<WorkspaceManifest>,

    /// If requested, contains the package manifest for the closest package in
    /// the workspace. `None` if there is no package manifest on the path to the
    /// workspace.
    /// If not requested this might still contain the package manifest stored in
    /// the same manifest as the workspace.
    pub package: Option<WithProvenance<PackageManifest>>,
}

/// An error that may occur when loading a discovered workspace directly from a
/// file.
#[derive(Debug, Error, Diagnostic)]
pub enum LoadManifestsError {
    #[error(transparent)]
    Io(#[from] std::io::Error),

    #[error(transparent)]
    #[diagnostic(transparent)]
    Toml(#[from] Box<WithSourceCode<TomlError, NamedSource<Arc<str>>>>),

    #[error(transparent)]
    #[diagnostic(transparent)]
    ProvenanceError(#[from] ProvenanceError),
}

impl Manifests {
    /// Constructs a new instance from a specific workspace manifest.
    pub fn from_workspace_manifest_path(
        workspace_manifest_path: PathBuf,
    ) -> Result<WithWarnings<Self, WarningWithSource>, LoadManifestsError> {
        let provenance = ManifestProvenance::from_path(workspace_manifest_path)?;
        let contents = provenance.read()?;
        Self::from_workspace_source(contents.into_inner().with_provenance(provenance))
    }

    /// Constructs a new instance from a specific workspace manifest that in
    /// memory.
    pub fn from_workspace_source<S: AsRef<str>>(
        WithProvenance {
            value: source,
            provenance,
        }: WithProvenance<S>,
    ) -> Result<WithWarnings<Self, WarningWithSource>, LoadManifestsError> {
        let build_source_code = || {
            NamedSource::new(
                provenance.path.to_string_lossy(),
                Arc::from(source.as_ref()),
            )
            .with_language(provenance.kind.language())
        };

        // Parse the TOML from the manifest.
        let mut toml = match toml_span::parse(source.as_ref()) {
            Ok(toml) => toml,
            Err(e) => {
                return Err(Box::new(WithSourceCode {
                    error: TomlError::from(e),
                    source: build_source_code(),
                })
                .into())
            }
        };

        // Parse the manifest as a workspace based on the type of manifest.
        let manifest_dir = provenance.path.parent().expect("a file must have a parent");
        let parsed_manifests = match provenance.kind {
            ManifestKind::Pixi => TomlManifest::deserialize(&mut toml)
                .map_err(TomlError::from)
                .and_then(|manifest| {
                    manifest.into_workspace_manifest(
                        ExternalWorkspaceProperties::default(),
                        Some(manifest_dir),
                    )
                }),
            ManifestKind::Pyproject => PyProjectManifest::deserialize(&mut toml)
                .map_err(TomlError::from)
                .and_then(|manifest| manifest.into_workspace_manifest(Some(manifest_dir))),
        };

        // Handle any errors that occurred during parsing.
        let (workspace_manifest, package_manifest, warnings) = match parsed_manifests {
            Ok(parsed_manifests) => parsed_manifests,
            Err(toml_error) => {
                return Err(Box::new(WithSourceCode {
                    error: toml_error,
                    source: build_source_code(),
                })
                .into())
            }
        };

        // Associate the warnings with the source code.
        let warnings = if warnings.is_empty() {
            vec![]
        } else {
            let source = build_source_code();
            warnings
                .into_iter()
                .map(|warning| WithSourceCode {
                    error: warning,
                    source: source.clone(),
                })
                .collect()
        };

        Ok(WithWarnings::from(Self {
            package: package_manifest
                .map(|package_manifest| WithProvenance::new(package_manifest, provenance.clone())),
            workspace: WithProvenance {
                provenance,
                value: workspace_manifest,
            },
        })
        .with_warnings(warnings))
    }
}

#[derive(Debug, Error, Diagnostic)]
pub enum ExplicitManifestError {
    #[error("could not find '{}'", .0.display())]
    MissingManifest(PathBuf),

    #[error(transparent)]
    InvalidManifest(ProvenanceError),

    #[error(transparent)]
    ParseVersionError(#[from] rattler_conda_types::ParseVersionError),

    /// The pixi version could not match the minimum requirement.
    #[error("workspace requires pixi {}, but I am {}", .requires_pixi, consts::PIXI_VERSION)]
    SelfVersionMatchError { requires_pixi: Version },
}

#[derive(Debug, Error, Diagnostic)]
pub enum WorkspaceDiscoveryError {
    #[error(transparent)]
    Io(#[from] std::io::Error),

    #[error(transparent)]
    #[diagnostic(transparent)]
    Toml(#[from] Box<WithSourceCode<TomlError, NamedSource<Arc<str>>>>),

    #[error(transparent)]
    #[diagnostic(transparent)]
    ExplicitManifestError(#[from] ExplicitManifestError),

    #[error("cannot canonicalize path '{1}' while searching for a manifest.")]
    Canonicalize(#[source] std::io::Error, PathBuf),
}

#[allow(clippy::large_enum_variant)]
enum EitherManifest {
    Pixi(TomlManifest),
    Pyproject(PyProjectManifest),
}

/// Defines where the search for the workspace should start.
#[derive(Debug, Clone)]
pub enum DiscoveryStart {
    /// Start the search from the given directory.
    ///
    /// This will search for a workspace manifest in the given directory and its
    /// parent directories.
    SearchRoot(PathBuf),

    /// Use the manifest file at the given path. Only search for a workspace if
    /// the specified manifest is not a workspace.
    ///
    /// This differs from specifying the parent directory of the manifest file
    /// in that it is also possible to specify a manifest that is not the
    /// default preferred format (e.g. `pyproject.toml`).
    ExplicitManifest(PathBuf),
}

impl DiscoveryStart {
    /// Returns the path of the directory or file to start the search from.
    pub fn root(&self) -> &Path {
        match self {
            DiscoveryStart::SearchRoot(root) => root.as_path(),
            DiscoveryStart::ExplicitManifest(manifest) => manifest.as_path(),
        }
    }
}

impl WorkspaceDiscoverer {
    /// Constructs a new instance from the current path.
    pub fn new(start: DiscoveryStart) -> Self {
        Self {
            start,
            discover_package: false,
        }
    }

    /// Also discover the package closest to the current directory.
    ///
    /// If set to `true`, the discoverer will also try to find the closest
    /// package manifest on the way to the workspace. Or if the workspace
    /// manifest also contains a package manifest it will be used as the closest
    /// package manifest.
    pub fn with_closest_package(self, discover_package: bool) -> Self {
        Self {
            discover_package,
            ..self
        }
    }

    /// Discover the workspace manifest in the directory tree.
    pub fn discover(
        self,
    ) -> Result<Option<WithWarnings<Manifests, WarningWithSource>>, WorkspaceDiscoveryError> {
        #[derive(Clone)]
        enum SearchPath {
            Explicit(PathBuf),
            Directory(PathBuf),
        }

        // Walk up the directory tree until we find a workspace manifest.
        let mut warnings = Vec::new();
        let mut closest_package_manifest = None;
        let mut next_search_path = match &self.start {
            DiscoveryStart::SearchRoot(root) => Some(SearchPath::Directory(
                dunce::canonicalize(root)
                    .map_err(|e| WorkspaceDiscoveryError::Canonicalize(e, root.clone()))?,
            )),
            DiscoveryStart::ExplicitManifest(manifest_path) => Some(SearchPath::Explicit(
                dunce::canonicalize(manifest_path)
                    .map_err(|e| WorkspaceDiscoveryError::Canonicalize(e, manifest_path.clone()))?,
            )),
        };
        while let Some(search_path) = next_search_path {
            let (next, provenance) = match search_path {
                SearchPath::Explicit(ref explicit) => {
                    if !explicit.exists() {
                        return Err(
                            ExplicitManifestError::MissingManifest(explicit.to_path_buf()).into(),
                        );
                    }
                    if explicit.is_file() {
                        let provenance = ManifestProvenance::from_path(explicit.clone())
                            .map_err(ExplicitManifestError::InvalidManifest)?;
                        let next_dir = explicit
                            .parent()
                            .expect("the manifest itself must have a parent directory")
                            .parent();
                        (next_dir.map(ToOwned::to_owned), Some(provenance))
                    } else {
                        let provenance = Self::provenance_from_dir(explicit).ok_or(
                            ExplicitManifestError::InvalidManifest(
                                ProvenanceError::UnrecognizedManifestFormat,
                            ),
                        )?;
                        tracing::info!(
                            "Found manifest in directory: {:?}, continuing further.",
                            provenance.path
                        );
                        (explicit.parent().map(ToOwned::to_owned), Some(provenance))
                    }
                }
                SearchPath::Directory(ref manifest_dir_path) => {
                    // Check if a pixi.toml file exists in the current directory.
                    let provenance = Self::provenance_from_dir(manifest_dir_path);
                    if provenance.is_some() {
                        tracing::info!(
                            "Found manifest in directory: {:?}, continuing further.",
                            manifest_dir_path
                        );
                    }
                    (
                        manifest_dir_path.parent().map(ToOwned::to_owned),
                        provenance,
                    )
                }
            };

            next_search_path = next.map(SearchPath::Directory);

            let Some(provenance) = provenance else {
                // If there is no manifest for the current search path, continue searching.
                continue;
            };

            // Read the contents of the manifest file.
            let contents = provenance.read()?.map(Arc::<str>::from);

            // Cheap check to see if the manifest contains a pixi section.
            if let ManifestSource::PyProjectToml(source) = &contents {
                if !source.contains("[tool.pixi")
                    && !matches!(search_path.clone(), SearchPath::Explicit(_))
                {
                    continue;
                }
            }

            let source = contents.into_named(provenance.absolute_path().to_string_lossy());

            // Parse the TOML from the manifest
            let mut toml = match toml_span::parse(source.inner()) {
                Ok(toml) => toml,
                Err(e) => {
                    return Err(Box::new(WithSourceCode {
                        error: TomlError::from(e),
                        source,
                    })
                    .into())
                }
            };

            // Parse the workspace manifest.
            let manifest_dir = provenance.path.parent().expect("a file must have a parent");
            let parsed_manifest = match provenance.kind {
                ManifestKind::Pixi => {
                    if closest_package_manifest.is_some() && toml.pointer("/workspace").is_none() {
                        // The manifest does not contain a workspace section, and we don't care
                        // about the package section.
                        continue;
                    }

                    // Parse as a pixi.toml manifest
                    let manifest = match TomlManifest::deserialize(&mut toml) {
                        Ok(manifest) => manifest,
                        Err(err) => {
                            return Err(Box::new(WithSourceCode {
                                error: TomlError::from(err),
                                source,
                            })
                            .into())
                        }
                    };

                    if manifest.has_workspace() {
                        // Parse the manifest as a workspace manifest if it contains a workspace
                        manifest.into_workspace_manifest(
                            ExternalWorkspaceProperties::default(),
                            Some(manifest_dir),
                        )
                    } else {
                        if self.discover_package {
                            // Otherwise store the manifest for later to parse as the closest
                            // package manifest.
                            closest_package_manifest = closest_package_manifest.or(Some((
                                EitherManifest::Pixi(manifest),
                                source,
                                provenance,
                            )));
                        }
                        continue;
                    }
                }
                ManifestKind::Pyproject => {
                    if closest_package_manifest.is_some()
                        && toml.pointer("/tool/pixi/workspace").is_none()
                    {
                        // The manifest does not contain a workspace section, and we don't care
                        // about the package section.
                        continue;
                    }

                    let manifest = match PyProjectManifest::deserialize(&mut toml) {
                        Ok(manifest) => manifest,
                        Err(err) => {
                            return Err(Box::new(WithSourceCode {
                                error: TomlError::from(err),
                                source,
                            })
                            .into())
                        }
                    };

                    if manifest.has_pixi_workspace() {
                        // Parse the manifest as a workspace manifest if it
                        // contains a workspace
                        manifest.into_workspace_manifest(Some(manifest_dir))
                    } else {
                        if self.discover_package {
                            // Otherwise store the manifest for later to parse as the closest
                            // package manifest.
                            closest_package_manifest = closest_package_manifest.or(Some((
                                EitherManifest::Pyproject(manifest),
                                source,
                                provenance,
                            )));
                        }
                        continue;
                    }
                }
            };

            let (workspace_manifest, package_manifest, workspace_warnings) = match parsed_manifest {
                Ok(parsed_manifest) => parsed_manifest,
                Err(error) => return Err(Box::new(WithSourceCode { error, source }).into()),
            };

            // Add the errors from the workspace manifest to the list of warnings.
            warnings.extend(
                workspace_warnings
                    .into_iter()
                    .map(|warning| WithSourceCode {
                        error: warning,
                        source: source.clone(),
                    }),
            );

            // Incorporate the workspace information into the package manifest.
            let closest_package_manifest = match closest_package_manifest {
                None => {
                    // If no package manifest is found on the way to the workspace manifest, we can
                    // use the package defined in the same manifest as the workspace itself.
                    package_manifest.map(|package_manifest| {
                        WithProvenance::new(package_manifest, provenance.clone())
                    })
                }
                Some((package_manifest, source, provenance)) => {
                    // Convert a found manifest into a package manifest using the workspace
                    // manifest.
                    let manifest_dir = provenance.path.parent().expect("a file must have a parent");
                    let package_manifest = match package_manifest {
                        EitherManifest::Pixi(manifest) => manifest.into_package_manifest(
                            ExternalPackageProperties::default(),
                            &workspace_manifest,
                            Some(manifest_dir),
                        ),
                        EitherManifest::Pyproject(manifest) => {
                            manifest.into_package_manifest(&workspace_manifest, Some(manifest_dir))
                        }
                    };

                    match package_manifest {
                        Ok((package_manifest, package_warnings)) => {
                            warnings.extend(package_warnings.into_iter().map(|warning| {
                                WithSourceCode {
                                    error: warning,
                                    source: source.clone(),
                                }
                            }));
                            Some(WithProvenance::new(package_manifest, provenance))
                        }
                        Err(error) => {
                            return Err(Box::new(WithSourceCode { error, source }).into());
                        }
                    }
                }
            };

            return Ok(Some(
                WithWarnings::from(Manifests {
                    workspace: WithProvenance::new(workspace_manifest, provenance),
                    package: closest_package_manifest,
                })
                .with_warnings(warnings),
            ));
        }

        Ok(None)
    }

    /// Discover the workspace manifest in a directory.
    fn provenance_from_dir(dir: &Path) -> Option<ManifestProvenance> {
        let pixi_toml_path = dir.join(consts::WORKSPACE_MANIFEST);
        let pyproject_toml_path = dir.join(consts::PYPROJECT_MANIFEST);
        if pixi_toml_path.is_file() {
            Some(ManifestProvenance::new(pixi_toml_path, ManifestKind::Pixi))
        } else if pyproject_toml_path.is_file() {
            Some(ManifestProvenance::new(
                pyproject_toml_path,
                ManifestKind::Pyproject,
            ))
        } else {
            None
        }
    }
}

#[cfg(test)]
mod test {
    use std::{fmt::Write, path::Path};

    use rstest::*;

    use super::*;
    use crate::utils::test_utils::format_diagnostic;

    #[rstest]
    #[case::root("")]
    #[case::empty("empty")]
    #[case::package_a("package_a")]
    #[case::package_b("package_a/package_b")]
    #[case::nested_workspace("nested-workspace")]
    #[case::nested_pyproject_workspace("nested-pyproject-workspace")]
    #[case::nested_pixi_project_in_nested_pyproject_workspace(
        "nested-pyproject-workspace/nested-pixi-project"
    )]
    #[case::nested_pyproject_in_nested_pyproject_workspace(
        "nested-pyproject-workspace/nested-pyproject"
    )]
    #[case::nested_non_pixi_pyproject_in_nested_pyproject_workspace(
        "nested-pyproject-workspace/nested-non-pixi-pyproject"
    )]
    #[case::non_pixi_build("non-pixi-build")]
    #[case::non_pixi_build_project("non-pixi-build/project")]
    fn test_workspace_discoverer(#[case] subdir: &str) {
        let test_data_root = dunce::canonicalize(
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../tests/data/workspace-discovery"),
        )
        .unwrap();

        let snapshot =
            match WorkspaceDiscoverer::new(DiscoveryStart::SearchRoot(test_data_root.join(subdir)))
                .with_closest_package(true)
                .discover()
            {
                Ok(None) => "Not found!".to_owned(),
                Ok(Some(WithWarnings {
                    value: discovered, ..
                })) => {
                    let rel_path =
                        pathdiff::diff_paths(&discovered.workspace.provenance.path, test_data_root)
                            .unwrap_or(discovered.workspace.provenance.path);
                    let mut snapshot = String::new();
                    writeln!(
                        &mut snapshot,
                        "Discovered workspace at: {}\n- Name: {}",
                        rel_path.display().to_string().replace("\\", "/"),
                        &discovered.workspace.value.workspace.name
                    )
                    .unwrap();

                    if let Some(package) = &discovered.package {
                        writeln!(
                            &mut snapshot,
                            "Package: {} @ {}",
                            &package.value.package.name, &package.value.package.version,
                        )
                        .unwrap();
                    }

                    snapshot
                }
                Err(e) => format_diagnostic(&e),
            };

        insta::with_settings!({
            snapshot_suffix => subdir.replace("/", "_"),
        }, {
            insta::assert_snapshot!(snapshot);
        });
    }

    #[rstest]
    #[case::root("")]
    #[case::pixi("pixi.toml")]
    #[case::empty("empty")]
    #[case::package_specific("package_a/pixi.toml")]
    fn test_explicit_workspace_discoverer(#[case] subdir: &str) {
        let test_data_root = dunce::canonicalize(
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../tests/data/workspace-discovery"),
        )
        .unwrap();

        let snapshot = match WorkspaceDiscoverer::new(DiscoveryStart::ExplicitManifest(
            test_data_root.join(subdir),
        ))
        .with_closest_package(true)
        .discover()
        {
            Ok(None) => "Not found!".to_owned(),
            Ok(Some(WithWarnings {
                value: discovered, ..
            })) => {
                let rel_path =
                    pathdiff::diff_paths(&discovered.workspace.provenance.path, test_data_root)
                        .unwrap_or(discovered.workspace.provenance.path);

                let mut snapshot = String::new();
                writeln!(
                    &mut snapshot,
                    "Discovered workspace at: {}\n- Name: {}",
                    rel_path.display().to_string().replace("\\", "/"),
                    &discovered.workspace.value.workspace.name
                )
                .unwrap();

                if let Some(package) = &discovered.package {
                    writeln!(
                        &mut snapshot,
                        "Package: {} @ {}",
                        &package.value.package.name, &package.value.package.version,
                    )
                    .unwrap();
                }

                snapshot
            }
            Err(e) => format_diagnostic(&e),
        };

        insta::with_settings!({
            snapshot_suffix => subdir.replace("/", "_"),
        }, {
            insta::assert_snapshot!(snapshot);
        });
    }

    #[test]
    fn test_non_existing_discovery() {
        // Split from the previous rstests, to avoid insta snapshot path conflicts in the error.
        let test_data_root = dunce::canonicalize(
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../tests/data/workspace-discovery"),
        )
        .unwrap();

        let err = WorkspaceDiscoverer::new(DiscoveryStart::SearchRoot(
            test_data_root.join("non-existing"),
        ))
        .with_closest_package(true)
        .discover()
        .expect_err("Expected an error");

        assert!(matches!(err, WorkspaceDiscoveryError::Canonicalize(_, _)));
    }
}
