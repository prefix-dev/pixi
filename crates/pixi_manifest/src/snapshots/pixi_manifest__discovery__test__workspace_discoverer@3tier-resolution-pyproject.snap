---
source: crates/pixi_manifest/src/discovery.rs
expression: snapshot
---
  × the workspace does not define a 'version'
    ╭─[<CARGO_ROOT>/tests/data/workspace-discovery/3tier-resolution-pyproject/pyproject.toml:10:11]
  9 │ name = "explicit-package-name"     # Tier 1: Direct value
 10 │ version = { workspace = true }     # Tier 2: Workspace inheritance
    ·           ────────────────────
 11 │ # authors not specified - should use tier 3 project defaults
    ╰────
